[AERIE-1674] Persist Scheduling Results

kroffo opened the pull request:
* **Tickets addressed:** AERIE-1674
* **Review:** By commit  <!-- Choose from: "by commit", "by file" -->
* **Merge strategy:** Merge (no squash)  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description
This branch provides the implementation of the scheduler-server's `PostgresResultsCellRepository`. The commits on this branch define the Postgres actions that will be used by the repository, and in the final commit the actions are hooked up tot he repository. Do note that some commits have descriptions explaining the motivation for their changes that reviewers may find useful.

## Verification
Each commit was tested by running `./gradlew classes` to make sure all commits compile. `./gradlew test` was also run at the end to ensure that no tests were broken. I also made sure the database initializes properly, since the scheduling_request table was modified.

As for testing the postgres actions themselves, I haven't run them since the scheduling agent is still in development.

## Documentation
No documentation updates needed. These features are in development, and not yet documented.

## Future work
Once the scheduler agent is hooked up we should do some integration testing and make sure a scheduling request can be created and run through our system, even if we have to manually insert the data into the database since the UI isn't hooked up yet. As careful as we try to be, it's difficult to be sure our code that interacts with the database is accurate without actually testing. Since none of this is in use yet, though, it shouldn't be a big deal if there's any issues to fix.

kroffo committed:
Implement postgres actions for getting/posting scheduling analysis

goetzrrGit commented on deployment/postgres-init-db/sql/scheduler/init.sql:
```diff
@@ -2,5 +2,15 @@
 
```
Just a note that @patkenneally PR changes the location of each service's SQL schema. Everything will be eventually moved to:
`<your-gradle-subproject>/sql/<table-name>
`

goetzrrGit replied:
And I just realized this is drafted :P

kroffo replied:
If that PR gets merged first I'll move the files 👍🏻

kroffo replied:
@goetzrrGit This is now addressed 👍🏻

kroffo committed:
Track only static data in scheduling PostgresResultsCell

A RequestRecord provides the state of a scheduling_request row at the
time it is requested. A PostgresResultCell's data is likely to persist
long after its creation, and thus the data in the RequestRecord it was
created from may very well be updated before it is used again. To
avoid keeping around stale data, we only keep references to the data
that identifies the request and its associated analysis, which should
never change.

kroffo committed:
Make scheduling requests cancellable

kroffo committed:
Provide all request data in RequestRecord

The RequestRecord is a record of a database table, and thus should
contain all columns of data. When originally written, I had unclear
boundaries between postgres data records and scheduler-server models,
and this helps make that boundary more clear. A follow-up commit will
remove the use of a RequestRecord in the PostgresResultsCell,
replacing with only the static data that should not change.

kroffo committed:
Implement action to update scheudling request state

kroffo committed:
Connect PostgresResultsCell to actions and thus the scheduler database

kroffo commented:
Force-pushed rebase off new develop to incorporate the SQL file location changes. No conflicts arose as git correctly applied the update to the scheduling_request table to the moved file.

kroffo commented:
Force-pushed a rebase off develop that also fixes the Error messages thrown by the cancel and state setting actions (they referred to the dataset table from merlin-server).

kroffo committed:
Fix SQL syntax and remove unused import

This file was created in a previous PR that was already merged, so
this update gets its own commit.

pcrosemurgy commented on scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/remotes/postgres/PostgresResultsCellRepository.java:
```diff
@@ -96,53 +108,185 @@ private static RequestRecord createRequest(
+          case "incomplete" -> new ResultsProtocol.State.Incomplete();
+          case "failed" -> new ResultsProtocol.State.Failed(request.failureReason());
+          case "success" -> new ResultsProtocol.State.Success(getResults(connection, request.analysisId()));
+          default -> throw new Error(String.format("Unexpected scheduling request state: %s", request.status()));
```
I'd be nice if `ResultRecord.status` was an `enum` type. Within the get request action you could parse the string. Just worries me since the `String status` does not offer compile-type guarantees and these strings are used in more than one place.

kroffo replied:
I would rather not parse the string in the action, because I believe the actions are merely for data retrieval. They provide the data that's in the database, but don't operate on it. The interpretation of that data is the responsibility of the relevant postgres repository, and I believe this boundary is beneficial to us (please let me know if your thoughts differ, I only just settled on my understanding of this boundary last week, and am always happy to hear other ideas/suggestions).

Now that said, I agree that having a switch statement on these string values feels less than ideal, but the string-to-State conversion is entirely contained in the `PostgresResultsCellRepository` in both directions*. In other words, the serialization/deserialization using strings for the State of a scheduling request is no more than an implementation detail of the `PostgresResultsCellRepository`.

`ResultsProtocol.State` is an interface (which could be made sealed) with three implementations, each with their own unique associated data. We can't do that with an enum, since all three variants have to have the same structure. We _could_ explore ways to move the strings into the interface implementations themselves, but I think this would cause more complexity than it's worth.

*the `"incomplete"` state is never explicitly set, but is the default value for new rows  added to the `scheduling_request` table

pcrosemurgy replied:
I'm cool with whatever you decide on here since at the end of the day it's fairly benign but thought I should give my 2 cents:

>I would rather not parse the string in the action, because I believe the actions are merely for data retrieval. They provide the data that's in the database, but don't operate on it. 

That's generally true but the boundary is not very clear for very discrete types like enums or ID records. For things like `PlanId` we already do parsing within actions, so I don't think an `incomplete`/`failed`/`success` parse is much of a stretch.

> In other words, the serialization/deserialization using strings for the State of a scheduling request is no more than an implementation detail of the PostgresResultsCellRepository.

Agreed, it is an implementation detail but the issue I see is exactly that: the implementation is error prone because we're relying on a custom and case-sensitive string in more than 1 place in the same class. `succeedRequest()` for example:
https://github.com/NASA-AMMOS/aerie/pull/52/files/edd39ea3767ee578e8c0165bb3ffeaddd90f688d#diff-cc4e1dc19800e82575716e95ec48d6a045df8df9af90ab257d7f56dd9293b829R159

>ResultsProtocol.State is an interface (which could be made sealed) with three implementations, each with their own unique associated data. We can't do that with an enum, since all three variants have to have the same structure.

This can be solved with double-dispatch (visitor pattern or `sealed` & `instanceof`).

kroffo replied:
Regarding the boundary between the server code and postgres action code, you're right, in the merlin-server, we build/parse `PlanId` records in postgres actions. We shouldn't. If you look at the new actions for the scheduler-server here, you'll see that I took care to avoid this, because I feel that we should keep our server models and postgres models separate. There are places where it almost feels silly, but keeping this boundary firm allows us to examine either side of it without having to keep the other in mind and reduces coupling between the server code and postgres code to a single location (the repository class itself). I do think if time permits it would actually be worth refactoring the merlin-server postgres actions to conform to this boundary as well.

As for the `ResultProtocol.State` implementation, I'm happy to explore other implementations that enable a more reliable postgres repository implementation, so I'll see what things would look like based on your suggestions and get back to you on that part.

pcrosemurgy replied:
Cool, and completely with you on the `PlanId` parsing stuff.

kroffo replied:
@pcrosemurgy Here's the thing about the `ResultsProtocol.State` problem. We need to go from string value to State instance and vice-versa. We could _get_ strings from State objects by adding a method to the State interface, but we can't go in the other direction without basically having a switch statement as we do now. And the fact that we need to compute results for a success variant, or a failure reason for a failed variant makes it even less clear how we would do such a thing.

mattdailis replied:
Passing thought: we should consider defining the valid values as a postgres enum, so the database will reject any strings that aren't valid.

kroffo replied:
Ya know, I actually explored that way back when working on merlin-server's simulation_dataset table, but I was still so new to postgres I didn't feel confident in what I was doing and gave up 😅

kroffo replied:
@pcrosemurgy Check out the latest commit on top of this branch. I implemented a Status enum in the RequestRecord to hide the string-comparison details away in a dedicated spot. This was based off some work done with @patkenneally and @mattdailis on the merlin-server yesterday. If desired, I can squash this commit into the relevant commits on the rest of this branch as well.

kroffo replied:
Furthermore, I've added a postgres enum `status_t` for the request status, so the only values it can ever take are the ones we allow (else postgres will give an error).

pcrosemurgy replied:
Thanks a lot, this looks great :)

pcrosemurgy commented:
Will look through more next week but just wanted to say I really appreciate how consistent you are with your DB actions.

kroffo commented:
> Will look through more next week but just wanted to say I really appreciate how consistent you are with your DB actions.

Konsistency is key!

camargo commented:
> Once the scheduler agent is hooked up we should do some integration testing and make sure a scheduling request can be created and run through our system, even if we have to manually insert the data into the database since the UI isn't hooked up yet. 

Can you please share with me (Wiki page maybe?) the GQL requests you make to insert the data when you run these manual integration tests? It will help me move on the UI more quickly if I have some concrete examples to play with.

mattdailis commented on scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/remotes/postgres/GetSatisfyingActivitiesAction.java:
```diff
@@ -0,0 +1,50 @@
+      final var goalId = new GoalId(resultSet.getLong("goal_id"));
+      final var activityId = new ActivityInstanceId(resultSet.getLong("activity_id"));
+
+      if (!satisfyingActivities.containsKey(goalId)) satisfyingActivities.put(goalId, new ArrayList<>());
```
This works just fine, but may be worth considering the "computeIfAbsent" method on maps.

Twisol replied:
For the particular use here (create a list and add to it), `computeIfAbsent` is truly lovely, since it returns the value whether you had to add it or not.

```java
      satisfyingActivities
        .computeIfAbsent(goalId, () -> new ArrayList<>())
        .add(activityId);
```

kroffo replied:
Sure thing, I never think to use these functions, thanks for the heads up, I'll try to keep them in mind!

mattdailis commented on scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/remotes/postgres/PostGoalSatisfactionAction.java:
```diff
@@ -0,0 +1,42 @@
+import java.sql.Statement;
+import java.util.Map;
+
+/*package-local*/ final class PostGoalSatisfactionAction implements AutoCloseable {
```
Did we ever reach a consensus on "Post" vs "Create" nomenclature? I think we had discussed renaming everything to Create in merlin-server, but had never gotten around to it.

EDIT: Or was it Insert?

kroffo replied:
We have a mix, and the reason seems to be that I naturally went for `create` when creating a new entity, such as creating a scheduling request, and `post` when inserting data that is associated with an existing thing, such as when posting scheduling results, which are associated with an existing request. Technically these "post" actions are still creating rows of data, and technically the "create" actions are posting data as well. I'm not sure if the distinction is actually beneficial, especially since it's undocumented (just in my head, I suppose). I'm happy to make things more consistent if we agree on an approach. I do feel that `insert` is the neutral ground between `create` and `post` here.

Twisol replied:
FWIW, "post" feels more like an HTTP / REST verb than a database / Postgres verb, though maybe that's just me. "insert" feels more generally DB-y. The verb "create" feels closer to the domain for some cases, but since this Action class is an implementation detail of a Repository (being package-private), maybe "create" makes more sense for the method on the Repository, and "insert" makes more sense for the low-level actions.

My two cents 😅

kroffo replied:
It seems that we generally lean toward "insert" here, so I've gone ahead and changed the "Post" actions over to "Insert" actions.

mattdailis reviewed (approved):
Looks good to me - I have no specific changes to request, just a few comments expressing some passing thoughts that should not block merge, especially since there are some tasks that will depend on this PR very soon.

mattdailis commented:
I noticed that `LocalSpecificationService` still throws UnsupportedOperationExceptions. It's not strictly speaking part of this PR, since it's not part of "Scheduling Results", but I just want to highlight it since I personally forgot about it until now.

https://github.com/NASA-AMMOS/aerie/blob/edd39ea3767ee578e8c0165bb3ffeaddd90f688d/scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/LocalSpecificationService.java#L8-L23

mattdailis commented on scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/remotes/postgres/GetRequestAction.java:
```diff
@@ -12,11 +10,14 @@
+      r.canceled
+    from scheduling_request as r
+    where
+      r.specification_id = ? and
```
🔻 Oop! We should make a call once and for all between `specification` and `spec`. This column name does not exist.

(My vote is to use the full word everywhere)

mattdailis replied:
There's a similar issue in `GetSpecificationAction`, which refers to the table `scheduling_specification`, which in reality is called `scheduling_spec`.

https://github.com/NASA-AMMOS/aerie/blob/edd39ea3767ee578e8c0165bb3ffeaddd90f688d/scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/remotes/postgres/GetSpecificationAction.java#L26

kroffo replied:
I'm going to probably go with specification for all except low-scope things like local variables, or cases of `as spec` for database tables in queries as above. I'll update you here once I've addressed this.

kroffo replied:
Alright, I added a commit to the start of this branch changing all instances of `spec` to `specification` in the database files. I stopped at every commit along the way in that rebase to make sure all postgres actions reflected this change.

kroffo replied:
I'll take the spaceship as a sign this is good to go, especially since the issue has been addressed, and Matt reached out to me to ask when I'm going to merge.

mattdailis commented on scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/remotes/postgres/SetRequestStateAction.java:
```diff
@@ -0,0 +1,46 @@
+  private static final @Language("SQL") String sql = """
+    update scheduling_request
+      set
+        state = ?,
```
```suggestion
        status = ?,
```

kroffo replied:
Nice catch!

mattdailis reviewed (changes requested):
One more small column name mismatch

kroffo commented:
> > Once the scheduler agent is hooked up we should do some integration testing and make sure a scheduling request can be created and run through our system, even if we have to manually insert the data into the database since the UI isn't hooked up yet.
> 
> Can you please share with me (Wiki page maybe?) the GQL requests you make to insert the data when you run these manual integration tests? It will help me move on the UI more quickly if I have some concrete examples to play with.

I can figure some out for you if you would like, but I haven't been using GraphQL for testing. I've been putting data into the database with TablePlus for now. Typically I'll add a few rows to scheduling_goal, then create a row of scheduling_spec, and add a row to scheduling_spec_goals for each scheduling_goal to associate the goals with the spec.

dyst5422 commented:
Sounds like a great opportunity to flex some more db-tests in.

Sent from my iPhone

On Feb 23, 2022, at 10:51, Kenny Roffo ***@***.***> wrote:

﻿

Once the scheduler agent is hooked up we should do some integration testing and make sure a scheduling request can be created and run through our system, even if we have to manually insert the data into the database since the UI isn't hooked up yet.

Can you please share with me (Wiki page maybe?) the GQL requests you make to insert the data when you run these manual integration tests? It will help me move on the UI more quickly if I have some concrete examples to play with.

I can figure some out for you if you would like, but I haven't been using GraphQL for testing. I've been putting data into the database with TablePlus for now. Typically I'll add a few rows to scheduling_goal, then create a row of scheduling_spec, and add a row to scheduling_spec_goals for each scheduling_goal to associate the goals with the spec.

—
Reply to this email directly, view it on GitHub<https://urldefense.us/v3/__https://github.com/NASA-AMMOS/aerie/pull/52*issuecomment-1049102009__;Iw!!PvBDto6Hs4WbVuu7!dZo4T8C4PKzVFDRo2eHswGEefyBuAQW5CUvV9Y7-6L3R5kcbQGL7fW3qW3HOV5SOaQqXX88-5FG2$>, or unsubscribe<https://urldefense.us/v3/__https://github.com/notifications/unsubscribe-auth/ADFK2Q4XGQCULCPRLFX2D6LU4UUBJANCNFSM5OV7VJZQ__;!!PvBDto6Hs4WbVuu7!dZo4T8C4PKzVFDRo2eHswGEefyBuAQW5CUvV9Y7-6L3R5kcbQGL7fW3qW3HOV5SOaQqXX64LZf6u$>.
You are receiving this because your review was requested.Message ID: ***@***.***>

mattdailis commented:
> > Can you please share with me (Wiki page maybe?) the GQL requests you make to insert the data when you run these manual integration tests? It will help me move on the UI more quickly if I have some concrete examples to play with.
> 
> I can figure some out for you if you would like, but I haven't been using GraphQL for testing. I've been putting data into the database with TablePlus for now. Typically I'll add a few rows to scheduling_goal, then create a row of scheduling_spec, and add a row to scheduling_spec_goals for each scheduling_goal to associate the goals with the spec.

I'm collecting a few graphql queries as I try to get an end to end test going to make sure I didn't screw up jar loading in #54 . I'd be happy to share those out once they're in a usable state.

mattdailis commented on scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/remotes/postgres/GetGoalSatisfactionAction.java:
```diff
@@ -0,0 +1,47 @@
+    select
+      goal.goal_id,
+      goal.satisfied
+    from schedulign_goal_analysis as goal
```
```suggestion
    from scheduling_goal_analysis as goal
```

kroffo replied:
Are you having trouble finding time to hangout with friends? Does it constantly seem like every time you have a free moment everyone you know is busy? If you're tired of this problem, then it's time to align! Align your schedules with SCHEDULIGN!

Twisol replied:
report spam

unsubscribe

kroffo replied:
Fixed!

dyst5422 commented:
These are the threads I came here for.

Sent from my iPhone

On Feb 23, 2022, at 14:27, Jonathan Castello ***@***.***> wrote:

﻿

@Twisol commented on this pull request.

________________________________

In scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/remotes/postgres/GetGoalSatisfactionAction.java<https://urldefense.us/v3/__https://github.com/NASA-AMMOS/aerie/pull/52*discussion_r813379717__;Iw!!PvBDto6Hs4WbVuu7!cTiA7yKufMbgGrzIpR80GEQ2NNmYSFdQVREcrg1K5kv45BEYxqyK4xj3YqAvz9ivk0KVI6j3xzdj$>:

> +import gov.nasa.jpl.aerie.scheduler.server.models.GoalId;
+import gov.nasa.jpl.aerie.scheduler.server.services.ScheduleResults;
+import org.intellij.lang.annotations.Language;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.Map;
+
+/*package-local*/ final class GetGoalSatisfactionAction implements AutoCloseable {
+  private final static @Language("SQL") String sql = """
+    select
+      goal.goal_id,
+      goal.satisfied
+    from schedulign_goal_analysis as goal


report spam

unsubscribe

—
Reply to this email directly, view it on GitHub<https://urldefense.us/v3/__https://github.com/NASA-AMMOS/aerie/pull/52*discussion_r813379717__;Iw!!PvBDto6Hs4WbVuu7!cTiA7yKufMbgGrzIpR80GEQ2NNmYSFdQVREcrg1K5kv45BEYxqyK4xj3YqAvz9ivk0KVI6j3xzdj$>, or unsubscribe<https://urldefense.us/v3/__https://github.com/notifications/unsubscribe-auth/ADFK2Q2U237DWOPQ4OUHF7DU4VNOFANCNFSM5OV7VJZQ__;!!PvBDto6Hs4WbVuu7!cTiA7yKufMbgGrzIpR80GEQ2NNmYSFdQVREcrg1K5kv45BEYxqyK4xj3YqAvz9ivk0KVI4ejXdrN$>.
You are receiving this because your review was requested.Message ID: ***@***.***>

kroffo committed:
Remove scheduling_rule table

The scheduling_rule SQL table was renamed to scheduling_goal in
AERIE-1672, but in AERIE-1666 all SQL files that existed on the
develop branch at that time were relocated. When git resolved the
resulting conflict, it considered the relocated scheduling_rule an
entirely new file, so its existence was not a conflict with the
removal of the original scheduling_rule file.

kroffo committed:
Hook up LocalSpecificationService to its repository

kroffo commented:
> I noticed that `LocalSpecificationService` still throws UnsupportedOperationExceptions. It's not strictly speaking part of this PR, since it's not part of "Scheduling Results", but I just want to highlight it since I personally forgot about it until now.
> 
> https://github.com/NASA-AMMOS/aerie/blob/edd39ea3767ee578e8c0165bb3ffeaddd90f688d/scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/LocalSpecificationService.java#L8-L23

That needed to be implemented, so I added a commit doing so.

kroffo committed:
Rename spec -> specification in scheduler db tables

In the interests of consistency and clarity, we rename instances of
"spec" to the full name "specification" among the database tables.

kroffo committed:
Hide request status strings behind an enum

Using an enum for request status, we can avoid string comparisons in
the results cell repository and other locations that needn't be overly
concerned with the representation of request stati.

kroffo commented:
Alright, after this morning I've addressed all comments. There's some new commits at both the front and back of this branch, so please check them out! The changes to existing commits were either directly related to a comment above, or an update to make the commit consistent with one of the additional commits.

kroffo committed:
Constrain scheduling_request status field with a postgres enum

mattdailis commented on scheduler-server/sql/scheduler/init.sql:
```diff
@@ -5,8 +5,8 @@ begin;
   \ir tables/scheduling_template_goals.sql
-  \ir tables/scheduling_spec.sql
-  \ir tables/scheduling_spec_goals.sql
+  \ir tables/scheduling_specification.sql
```
Glad to see this renaming. We still need to update the public_scheduling_spec.yaml hasura metadata

mattdailis replied:
Update: I've finished my test, and I did not find any more errors in files relevant to this PR. I'll make a separate hotfix PR for the other errors I found 👍

kroffo replied:
Nice catch, should be fixed 👍🏻

mattdailis commented on scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/remotes/postgres/GetSpecificationGoalsAction.java:
```diff
@@ -1,6 +1,5 @@
 package gov.nasa.jpl.aerie.scheduler.server.remotes.postgres;
 
-import gov.nasa.jpl.aerie.scheduler.server.models.SpecificationId;
```
The SQL in this file is invalid - there should be no comma after the closing parenthesis, and the table is called `scheduling_goal` singular.

kroffo replied:
Fixed this in the "Remove unused import" commit (and updated the commit message)

mattdailis reviewed (changes requested):
I only got about halfway through my test, but I'll give you these comments now so as not to keep you waiting

mattdailis reviewed (approved):
Great work, Kenny 🚀 , thanks for addressing my concerns.
